core/src/main/java/com/amplifyframework/predictions/models/Feature.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.predictions.models;
+
+import androidx.annotation.NonNull;
+
+import java.util.Objects;
+import java.util.UUID;
+
+/**
+ * A generic class to hold information about an inferred
+ * feature and the confidence score for inference.
+ * @param <T> the feature type
+ */
+public abstract class Feature<T> implements Comparable<Feature<T>> {
+    private final String id;
+    private final T feature;
+    private final float confidence;
+
+    Feature(Builder<?, ? extends Feature<T>, T> builder) {
+        this.id = builder.getId();
+        this.feature = builder.getFeature();
+        this.confidence = builder.getConfidence();
+    }
+
+    /**
+     * Gets the type name of feature.
+     * @return the feature type
+     */
+    @NonNull
+    public abstract String getType();
+
+    /**
+     * Gets the unique ID assigned to this result.
+     * @return unique ID
+     */
+    @NonNull
+    public final String getId() {
+        return id;
+    }
+
+    /**
+     * Gets the detected value for this feature.
+     * @return the detected result
+     */
+    @NonNull
+    public final T getFeature() {
+        return feature;
+    }
+
+    /**
+     * Gets the confidence score for this detection result.
+     * The confidence score is a percentage value between
+     * 0 and 100.
+     * @return the confidence score
+     */
+    public final float getConfidence() {
+        return confidence;
+    }
+
+    /**
+     * Compares an feature to another. The features are
+     * sorted by their types' alphabetic order, and by
+     * decreasing order of their confidence score for those
+     * within the same type.
```

Comment:
Would it make sense to sort only by confidence? (Just a question.)

Reply:
It is sorted only by confidence as long as you are comparing two features of same type. It wouldn't really make sense to compare the confidence rating of sentiment detection with say... language detection. I was considering throwing a runtime exception if the feature type mismatches, but I just left it like this.

Confidence sorting would make sense if an operation returns a list of results (for example, if text interpretation returned an output like `[70% positive, 10% negative, 20% neutral],` (just an example), then a user can sort this list to find the predominant sentiment (first element of sorted list) or something like that.